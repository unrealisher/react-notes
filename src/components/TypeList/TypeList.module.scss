.list {
  display: inline-flex;
  border-radius: 15px;
  overflow: hidden;
  border: 1px solid rgba(218, 220, 224, 0.5);
  background-color: white;
}

.item {
  position: relative;

  &:not(:last-child)::after {
    content: "";
    position: absolute;
    width: 2px;
    height: 100%;
    top: 0;
    left: 100%;
    background-color: rgb(163, 163, 163);
    z-index: 2;
  }
}

.label {
  cursor: pointer;
  padding: 3px 8px;
}

.radio:checked + .label {
  background-color: #fc0;
}
