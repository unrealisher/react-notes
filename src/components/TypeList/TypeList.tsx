import React from "react";
import cn from "classnames";

import styles from "./TypeList.module.scss";

interface IItem {
  text: string;
  value: string;
}

interface IProps {
  wrapper?: string;
  type: string;
  name: string;
  onTypeChange: Function;
  items: IItem[];
}

const TypeList = (props: IProps): JSX.Element => {
  const { wrapper, type, onTypeChange, items, name } = props;
  const getItems = (): JSX.Element[] => {
    return items.map(item => {
      return (
        <li key={item.value} className={styles.item}>
          <input
            id={item.value + name}
            className={styles.radio}
            type="radio"
            name={name}
            value={item.value}
            onChange={() => onTypeChange(item.value)}
            checked={type === item.value}
            hidden
          />
          <label className={styles.label} htmlFor={item.value + name}>
            {item.text}
          </label>
        </li>
      );
    });
  };
  return (
    <ul className={cn(styles.list, wrapper ? wrapper : undefined)}>
      {getItems()}
    </ul>
  );
};

export default TypeList;
