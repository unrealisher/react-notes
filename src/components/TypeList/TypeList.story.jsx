import React from "react";

import TypeList from "./TypeList";

import { storiesOf } from "@storybook/react";

import "./../../index.scss";

storiesOf("TypeList", module).add("стандарт", () => <TypeList />);
