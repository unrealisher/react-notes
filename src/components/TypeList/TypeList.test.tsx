import React from "react";
import { shallow, mount } from "enzyme";
import { TypeList } from "./TypeList";

describe("TypeList", () => {
  const type = "text";
  const name = "text";
  const onTypeChange = (): void => {};
  const items = [
    { text: "Список", value: "list" },
    { text: "Текст", value: "text" },
    { text: "Картинка", value: "image" }
  ];

  it("correct render", () => {
    const component = shallow(
      <TypeList
        type={type}
        name={name}
        onTypeChange={onTypeChange}
        items={items}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("onTypeChange calls when button is clicked", () => {
    const onTypeChange = jest.fn();
    const component = mount(
      <TypeList
        type={type}
        name={name}
        onTypeChange={onTypeChange}
        items={items}
      />
    );
    const radio = component.find(".radio");
    radio.forEach(item => item.simulate("change"));
    expect(onTypeChange).toBeCalledTimes(3);
  });
});
