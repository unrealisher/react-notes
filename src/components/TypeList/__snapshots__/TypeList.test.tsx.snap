// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeList correct render 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TypeList
    items={
      Array [
        Object {
          "text": "Список",
          "value": "list",
        },
        Object {
          "text": "Текст",
          "value": "text",
        },
        Object {
          "text": "Картинка",
          "value": "image",
        },
      ]
    }
    name="text"
    onTypeChange={[Function]}
    type="text"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <li
          className="item"
        >
          <input
            checked={false}
            className="radio"
            hidden={true}
            id="listtext"
            name="text"
            onChange={[Function]}
            type="radio"
            value="list"
          />
          <label
            className="label"
            htmlFor="listtext"
          >
            Список
          </label>
        </li>,
        <li
          className="item"
        >
          <input
            checked={true}
            className="radio"
            hidden={true}
            id="texttext"
            name="text"
            onChange={[Function]}
            type="radio"
            value="text"
          />
          <label
            className="label"
            htmlFor="texttext"
          >
            Текст
          </label>
        </li>,
        <li
          className="item"
        >
          <input
            checked={false}
            className="radio"
            hidden={true}
            id="imagetext"
            name="text"
            onChange={[Function]}
            type="radio"
            value="image"
          />
          <label
            className="label"
            htmlFor="imagetext"
          >
            Картинка
          </label>
        </li>,
      ],
      "className": "list",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "list",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              checked={false}
              className="radio"
              hidden={true}
              id="listtext"
              name="text"
              onChange={[Function]}
              type="radio"
              value="list"
            />,
            <label
              className="label"
              htmlFor="listtext"
            >
              Список
            </label>,
          ],
          "className": "item",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "checked": false,
              "className": "radio",
              "hidden": true,
              "id": "listtext",
              "name": "text",
              "onChange": [Function],
              "type": "radio",
              "value": "list",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Список",
              "className": "label",
              "htmlFor": "listtext",
            },
            "ref": null,
            "rendered": "Список",
            "type": "label",
          },
        ],
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "text",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              checked={true}
              className="radio"
              hidden={true}
              id="texttext"
              name="text"
              onChange={[Function]}
              type="radio"
              value="text"
            />,
            <label
              className="label"
              htmlFor="texttext"
            >
              Текст
            </label>,
          ],
          "className": "item",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "checked": true,
              "className": "radio",
              "hidden": true,
              "id": "texttext",
              "name": "text",
              "onChange": [Function],
              "type": "radio",
              "value": "text",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Текст",
              "className": "label",
              "htmlFor": "texttext",
            },
            "ref": null,
            "rendered": "Текст",
            "type": "label",
          },
        ],
        "type": "li",
      },
      Object {
        "instance": null,
        "key": "image",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              checked={false}
              className="radio"
              hidden={true}
              id="imagetext"
              name="text"
              onChange={[Function]}
              type="radio"
              value="image"
            />,
            <label
              className="label"
              htmlFor="imagetext"
            >
              Картинка
            </label>,
          ],
          "className": "item",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "checked": false,
              "className": "radio",
              "hidden": true,
              "id": "imagetext",
              "name": "text",
              "onChange": [Function],
              "type": "radio",
              "value": "image",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Картинка",
              "className": "label",
              "htmlFor": "imagetext",
            },
            "ref": null,
            "rendered": "Картинка",
            "type": "label",
          },
        ],
        "type": "li",
      },
    ],
    "type": "ul",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <li
            className="item"
          >
            <input
              checked={false}
              className="radio"
              hidden={true}
              id="listtext"
              name="text"
              onChange={[Function]}
              type="radio"
              value="list"
            />
            <label
              className="label"
              htmlFor="listtext"
            >
              Список
            </label>
          </li>,
          <li
            className="item"
          >
            <input
              checked={true}
              className="radio"
              hidden={true}
              id="texttext"
              name="text"
              onChange={[Function]}
              type="radio"
              value="text"
            />
            <label
              className="label"
              htmlFor="texttext"
            >
              Текст
            </label>
          </li>,
          <li
            className="item"
          >
            <input
              checked={false}
              className="radio"
              hidden={true}
              id="imagetext"
              name="text"
              onChange={[Function]}
              type="radio"
              value="image"
            />
            <label
              className="label"
              htmlFor="imagetext"
            >
              Картинка
            </label>
          </li>,
        ],
        "className": "list",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "list",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                checked={false}
                className="radio"
                hidden={true}
                id="listtext"
                name="text"
                onChange={[Function]}
                type="radio"
                value="list"
              />,
              <label
                className="label"
                htmlFor="listtext"
              >
                Список
              </label>,
            ],
            "className": "item",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "checked": false,
                "className": "radio",
                "hidden": true,
                "id": "listtext",
                "name": "text",
                "onChange": [Function],
                "type": "radio",
                "value": "list",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Список",
                "className": "label",
                "htmlFor": "listtext",
              },
              "ref": null,
              "rendered": "Список",
              "type": "label",
            },
          ],
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "text",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                checked={true}
                className="radio"
                hidden={true}
                id="texttext"
                name="text"
                onChange={[Function]}
                type="radio"
                value="text"
              />,
              <label
                className="label"
                htmlFor="texttext"
              >
                Текст
              </label>,
            ],
            "className": "item",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "checked": true,
                "className": "radio",
                "hidden": true,
                "id": "texttext",
                "name": "text",
                "onChange": [Function],
                "type": "radio",
                "value": "text",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Текст",
                "className": "label",
                "htmlFor": "texttext",
              },
              "ref": null,
              "rendered": "Текст",
              "type": "label",
            },
          ],
          "type": "li",
        },
        Object {
          "instance": null,
          "key": "image",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                checked={false}
                className="radio"
                hidden={true}
                id="imagetext"
                name="text"
                onChange={[Function]}
                type="radio"
                value="image"
              />,
              <label
                className="label"
                htmlFor="imagetext"
              >
                Картинка
              </label>,
            ],
            "className": "item",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "checked": false,
                "className": "radio",
                "hidden": true,
                "id": "imagetext",
                "name": "text",
                "onChange": [Function],
                "type": "radio",
                "value": "image",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Картинка",
                "className": "label",
                "htmlFor": "imagetext",
              },
              "ref": null,
              "rendered": "Картинка",
              "type": "label",
            },
          ],
          "type": "li",
        },
      ],
      "type": "ul",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
